lib/withOnyx.d.ts

```diff
@@ -104,8 +104,8 @@ type CollectionMapping<TComponentProps, TOnyxProps, TOnyxProp extends keyof TOny
     CollectionBaseMapping<TOnyxKey> &
     (
         | BaseMappingKey<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey, OnyxCollection<KeyValueMapping[TOnyxKey]>>
-        | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey>
-        | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey>
+        | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, Partial<KeyValueMapping[TOnyxKey]>, TOnyxKey>
+        | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, Partial<KeyValueMapping[TOnyxKey]>, TOnyxKey>
```

Comment:
```suggestion
        | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, Partial<KeyValueMapping[TOnyxKey]>, TOnyxKey>
        | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, Partial<KeyValueMapping[TOnyxKey]>, TOnyxKey>
```

@tienifr We have two issues here:
1. By changing `TOnyxProps[TOnyxProp]` to `Partial<KeyValueMapping[TOnyxKey]>` we lose all type equality check between the components' `OnyxProps` and `withOnyx` mapping, which means that we could type anything in the Onyx prop and the selector wouldn't be able to check if its return type matches with the corresponding Onyx prop.
2. Also, by doing `Partial<KeyValueMapping[TOnyxKey]>` you are limiting the selector's return to be an object with optional properties of that Onyx value, _which is not always the case_.
   
   When we have a combination of a collection key and a `selector`, the `withOnyx` will run `selector` function for each entry of that collection and build a collection object with the results of all `selector` executions, for example:

   ```tsx
   type ContactMethodDetailsPageOnyxProps = {
       policiesIDs: OnyxCollection<string>;
   };
   
   ...
   
   export default withOnyx<ContactMethodDetailsPageProps, ContactMethodDetailsPageOnyxProps>({
       policiesIDs: {
           key: ONYXKEYS.COLLECTION.POLICY,
           selector: (data) => data?.id ?? '', // this is valid and will make policies return a collection of ids
       },
   })(ContactMethodDetailsPage);
   ```

To improve these typings, my suggestion is:

1. Add this type in `lib/types.d.ts` which will make possible extracting the value of a Onyx collection:
   ```tsx
   ...
   type OnyxCollection<TOnyxValue> = OnyxEntry<Record<string, TOnyxValue | null>>;

   /**
    * Utility type to extract `TOnyxValue` from `OnyxCollection<TOnyxValue>`.
    */
   type ExtractOnyxCollectionValue<TOnyxCollection> = TOnyxCollection extends NonNullable<OnyxCollection<infer U>> ? U : never;
   ...
   ```
2. Use it in `CollectionMapping`:
   ```tsx
   /**
    * Represents the mapping options between an Onyx collection key without suffix and the component's prop with all its possibilities.
    */
   type CollectionMapping<TComponentProps, TOnyxProps, TOnyxProp extends keyof TOnyxProps, TOnyxKey extends CollectionKeyBase> = BaseMapping<TComponentProps, TOnyxProps> &
       CollectionBaseMapping<TOnyxKey> &
       (
           | BaseMappingKey<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey, OnyxCollection<KeyValueMapping[TOnyxKey]>>
           | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, ExtractOnyxCollectionValue<TOnyxProps[TOnyxProp]>, TOnyxKey>
           | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, ExtractOnyxCollectionValue<TOnyxProps[TOnyxProp]>, TOnyxKey>
       );
   ```

This way we can ensure the type equality checks will still work and get rid of these issues when using the `selector`. For your case [in this discussion](https://github.com/Expensify/App/pull/36713/files#discussion_r1502293136), final solution would be like this:

```tsx
type ContactMethodDetailsPageOnyxProps = {
    ...
    policies: OnyxCollection<Pick<Policy, 'id' | 'ownerAccountID'>>;
};

...

export default withOnyx<ContactMethodDetailsPageProps, ContactMethodDetailsPageOnyxProps>({
    ...
    policies: {
        key: ONYXKEYS.COLLECTION.POLICY,
        selector: (data) => ({id: data?.id ?? '', ownerAccountID: data?.ownerAccountID}),
    },
})(ContactMethodDetailsPage);
```